nixos.nix

```
@@ -90,13 +90,67 @@ in
         # Create a directory in persistent storage, so we can bind
         # mount it.
         mkDirCreationSnippet = persistentStoragePath: dir:
-          let
-            targetDir = concatPaths [ persistentStoragePath dir ];
-          in
           ''
-            if [[ ! -e "${targetDir}" ]]; then
-                mkdir -p "${targetDir}"
-            fi
+            # Given a source directory (/source) and a target directory
+            # (/target/foo/bar/bazz) we want to "clone" the target structure
+            # from source into the target, which is to say we want to get both
+            # /source/target/foo/bar/bazz and /target/foo/bar/bazz to exist
+            # in the filesystem. More concretely, we'd like to map
+            # /state/etc/ssh/example.key to /etc/ssh/example.key
+            #
+            # To do this we split the target's path into parts (target, foo,
+            # bar, bazz) and iterate over them while accumulating the path
+            # (/target/, /target/foo/, /target/foo/bar, ...) and for each of
+            # these incresingly qualified paths we:
+            #
+            #  1. Enforce /source/qualifiedPath and qualifiedPath exists
+            #  2. Copy the ownership of the source path into the target path
+            #  3. Copy the mode of the source path into the target path
```

"we want to get both /source/target/foo/bar/bazz [1] and /target/foo/bar/bazz to exist" -- shouldn't [1] already exist?

---

```suggestion
            # Given a source directory, /source, and a target directory,
            # (/target/foo/bar/bazz, we want to "clone" the target structure
            # from source into the target. Essentially, we want to get both
            # /source/target/foo/bar/bazz and /target/foo/bar/bazz to exist
            # in the filesystem. More concretely, we'd like to map
            # /state/etc/ssh/example.key to /etc/ssh/example.key
            #
            # To achieve this, we split the target's path into parts -- target, foo,
            # bar, bazz -- and iterate over them while accumulating the path
            # (/target/, /target/foo/, /target/foo/bar, and so on); then, for each of
            # these increasingly qualified paths we:
            #
            #   1. Ensure both /source/qualifiedPath and qualifiedPath exist
            #   2. Copy the ownership of the source path into the target path
            #   3. Copy the mode of the source path into the target path
```